### PR TITLE
feat: PEFT fine-tuning pipeline (LoRA/DoRA, multi-GPU) for TimesFM 2.5

### PR DESCRIPTION
**Pull request by darkpowerxo**
## Summary

Adds a full PEFT (Parameter-Efficient Fine-Tuning) pipeline for TimesFM 2.5 with LoRA and DoRA adapters, multi-GPU support via PyTorch DDP, and mixed-precision training. Also includes unit tests, bug fixes, documentation updates, and several cherry-picked community fixes.

### PEFT Pipeline (`peft/`)

```
peft/
├── __init__.py       ← public API exports
├── config.py         ← PEFTConfig dataclass (adapter type, rank, target modules, etc.)
├── adapters.py       ← LoRA/DoRA layers, inject/merge/save/load adapter weights
├── data.py           ← TimeSeriesDataset (sliding-window from arrays or DataFrames)
├── trainer.py        ← PEFTTrainer: DDP, AMP, cosine-warmup, early stopping, W&B logging
├── finetune.py       ← CLI entry-point (argparse)
├── finetune.sh       ← torchrun launch script with NUM_GPUS control
└── README.md         ← full documentation and usage examples
```

Key features:
- **LoRA and DoRA** adapters targeting `qkv_proj`, `output_proj`, and `ff_layer` modules
- **`num_adapter_layers`** parameter to freeze all but the last N transformer layers
- **Multi-GPU** training via `torchrun` and `DistributedDataParallel`
- **Mixed-precision** (FP16/BF16) via `torch.amp`
- **safetensors** adapter-only checkpoints (~4 MB at rank-4)
- **Gradient checkpointing** for memory-constrained setups

### Unit Tests (`tests/`)

58 tests covering configs, torch layers, torch utilities, and base utilities:

```
tests/
├── __init__.py
├── test_base_utils.py
├── test_configs.py
├── test_torch_layers.py
└── test_torch_utils.py
```

### Bug Fixes & Improvements

- **fix: per-input ridge regression** in `xreg_lib.py` (both `src/` and `v1/`) to prevent data leakage across batch items
- **fix: `batch_size` iteration** in `v1/src/timesfm/data_loader.py` when `permute=False`
- **fix: correct SKILL.md link** in README (`tree` → `blob` path)
- **fix: typo** `complied` → `compiled` in `configs.py`
- **docs: fix swapped `xreg_mode` descriptions** and `covaraites` typo in `timesfm_2p5_base.py`
- **refactor: `print` → `logging.info`** in `timesfm_2p5_torch.py`
- **ci: upgrade GitHub Actions** (`checkout` / `setup-python` v2 → v6)
- **docs: update README** — replace "under construction" section with completed feature checklist

## No breaking changes

All source-level modifications are backward-compatible:
- `configs.py`, `timesfm_2p5_base.py`: docstring/typo fixes only
- `timesfm_2p5_torch.py`: `print` → `logging.info`
- `xreg_lib.py`: per-input regression refactor preserves the existing 5-tuple return signature

## Testing

```
pytest tests/ -q
# 58 passed in 1.46s
```

All 58 tests pass. PEFT imports, `import timesfm`, and `ForecastConfig` all verified working.

## Incorporated community PRs

| PR | Description |
|----|-------------|
| #396 | fix: correct typo + replace print with logging |
| #394 | test: add unit tests for configs, torch layers, utils |
| #393 | fix: per-input ridge regression in xreg |
| #391 | fix: batch_size in v1 data_loader |
| #390 | fix: SKILL.md link in README |
| #367 | ci: upgrade GitHub Actions to v6 |
| #366 | docs: fix swapped xreg_mode descriptions |

### CODE REVIEW

**Comment by darkpowerxo** [comment]
@siriuz42 @kashif @heidmotron @sitic @abhidas 

Can we get an idea of how long would it take to have the pr approved? 
so i can start working on other PRs

**Comment by kashif** [comment]
its already supported see: https://github.com/huggingface/notebooks/blob/main/examples/timesfm2_5.ipynb

**Comment by darkpowerxo** [comment]
Thanks @kashif — you're absolutely right, and great work on the Transformers integration! I wasn't aware of your notebook when I opened this PR.

I've reworked the PR accordingly: the custom PEFT pipeline is gone. Instead, this now ships a lightweight Python script finetune_lora.py that uses the standard Transformers + PEFT workflow — same dataset, same LoRA config, same training API as your notebook. Think of it as a runnable, CLI-friendly version of [your example](https://github.com/huggingface/notebooks/blob/main/examples/timesfm2_5.ipynb) that lives alongside the other TimesFM examples in this repo.

The notebook is excellent for interactive exploration, but having a standalone script here makes fine-tuning more discoverable for users who start from this repo rather than the HuggingFace notebooks collection.

The rest of the PR includes bug fixes, CI upgrades, and documentation — all independent of the fine-tuning example. Happy to adjust anything based on your feedback.

**Comment by siriuz42** [comment]
Thanks @darkpowerxo! Working on it now.

**Comment by siriuz42** [comment]
Some initial comments:

1. Let's not merge #393. The batch behavior is intended. One can run batch_size=1 inference to not cross the examples.

2. Let's not merge #391. Let's keep the deprecated v1 code as is.

3. Let's not merge #390. It's intended to link to that page.

File: src/timesfm/timesfm_2p5/timesfm_2p5_torch.py [hunk 1]
```diff
@@ -85,7 +85,7 @@ def load_checkpoint(self, path: str, **kwargs):
     if "torch_compile" in kwargs:
       torch_compile = kwargs["torch_compile"]
     if torch_compile:
-      print("Compiling model...")
+      logging.info("Compiling model...")
```
Comment (siriuz42) [inline src/timesfm/timesfm_2p5/timesfm_2p5_torch.py:88]: Let's keep both. print can come handy in a notebook

**Comment by siriuz42** [comment]
@kashif Can we get a LGTM from you regarding the HF transformers part?

**Comment by darkpowerxo** [comment]
Thanks for the review @siriuz42!

Done — I've reverted both:

#393 (per-input ridge regression in xreg) — understood, batch behavior is intended and users can run [batch_size=1](vscode-file://vscode-app/c:/Program%20Files/Microsoft%20VS%20Code/41dd792b5e/resources/app/out/vs/code/electron-browser/workbench/workbench.html) for per-input semantics.
#390 (SKILL.md link) — reverted to the original directory link.
The remaining changes in this PR are:

Fine-tuning example ([finetuning](vscode-file://vscode-app/c:/Program%20Files/Microsoft%20VS%20Code/41dd792b5e/resources/app/out/vs/code/electron-browser/workbench/workbench.html)) — a CLI-friendly Python script using the standard HuggingFace Transformers + PEFT workflow (based on @kashif's notebook)
Unit tests ([tests](vscode-file://vscode-app/c:/Program%20Files/Microsoft%20VS%20Code/41dd792b5e/resources/app/out/vs/code/electron-browser/workbench/workbench.html)) for core layers, configs, and utilities
Bug fixes in configs and model code (e.g., missing freq attribute, batch dim assertions)
CI workflow upgrades (actions v6)
Documentation updates

File: src/timesfm/timesfm_2p5/timesfm_2p5_torch.py (hunk 1 above)
Comment (siriuz42) [reply src/timesfm/timesfm_2p5/timesfm_2p5_torch.py:88]: In hindsight let's merge first.

**Comment by siriuz42** [comment]
Thanks for the contribution, [darkpowerxo](https://github.com/darkpowerxo)!
